### PR TITLE
🪐 Support for JupyterLab 4.x

### PR DESCRIPTION
**Pull request by agoose77**
This PR will probably take a while. The new windowed notebook breaks things that will require some care to recover.

Simultaneously, I think we should think about how #102 fits into this PR at the same time.

Once this gets merged, it might be more straightforward for us to only support 4.x. I haven't had much experience in maintaining compatibility across multiple versions. The obvious way is to make a branch from `HEAD~1` after merging this PR, and bump the version in `main` to `2.0.0`. Then we can continue to make backports if needs be.

### CODE REVIEW

**Comment by fperez** [comment]
@agoose77 - I'm very happy to see this progress, feel free to ping me if at some point you want some testing/feedback I can help with. Happy to do it, as having this entire toolchain happily spinning on JL4 is important to me.  And huge thanks for the amazing work, once more :)

**Comment by agoose77** [comment]
@rowanc1 @fperez could you take this for a spin? I was seeing some local performance regressions in Firefox, but I can't reproduce them today. 

I'm aware of a double-render issue with executing inline expression cells; I plan to fix that in the coming week.

File: README.md [hunk 1]
```diff
@@ -115,14 +136,14 @@ The `jlpm` command is JupyterLab's pinned version of
 `yarn` or `npm` in lieu of `jlpm` below.
```
Comment (rowanc1) [inline README.md:136]: I don't think this is true anymore. I had to use the bundled yarn version that `jlpm install` uses.

File: package.json [hunk 2]
```diff
@@ -1,125 +1,127 @@
...
+    "@jupyterlab/rendermime": "^4.0.0",
+    "@jupyterlab/translation": "^4.0.0",
+    "@myst-theme/diagrams": "^0.2.6",
+    "@myst-theme/frontmatter": "^0.2.6",
+    "@myst-theme/providers": "^0.2.6",
+    "katex": "^0.15.2",
+    "myst-ext-card": "^0.0.7",
+    "myst-ext-grid": "^0.0.7",
+    "myst-ext-proof": "^0.0.2",
+    "myst-ext-tabs": "^0.0.7",
```
Comment (rowanc1) [inline package.json:74]: We can bring the new stuff forward pretty easy. If we are calling this the new "main" let me know, and I can rebase the changes over here.

File: src/TextModelProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+export interface ITextModel {
+  getSource(): string;
+  setSource(source: string): void;
+}
+
+type TextModelState = {
+  model?: ITextModel;
+};
+
+const TextModelContext = createContext<TextModelState | undefined>(undefined);
```
Comment (rowanc1) [inline src/TextModelProvider.tsx:12]: Does this mean that task lists work in the markdown preview as well?

**Comment by fperez** [comment]
Thanks @rowanc1! I was actually trying to set up a dev environment to test it, but I got mired in problems with conda/mamba... Argh... In any case, it seems like your round of input may give enough for now - I'll finish cleaning up my local setup so I can help with a round of testing after more updates.

If the yarn/jlpm fixes go in that would be great - I know so little about js/ts development that I'm likely to stumble hard on that if it doesn't work out of the box as per the README (which is what I was planning on blindly following).

**Review by rowanc1 — COMMENTED** [review]
A few of the cells in the example notebook have been removed. Is this intentional?

One of the matplotlib examples does actually fail now, this could be related to the double render, so I think we should fix that first!

I updated trust in #150.

File: README.md (hunk 1 above)
Comment (agoose77) [reply README.md:136]: Hmm, let's check the upstream extensions cookie-cutter (but in general I recall the version always being fairly important).

File: src/TextModelProvider.tsx (hunk 3 above)
Comment (agoose77) [reply src/TextModelProvider.tsx:12]: Yes & no. RenderMime isn't supposed to write back to the model IIRC, but we could probably do it if the model we're given is directly associated with the file content, i.e. not a copy. If this were the case, we would need to do a bit more here, though.

**Comment by parmentelat** [comment]
I am also in the process of putting together a jlab4-based env for next year courses starting in September

in order to try out this wip I was expecting to be able to do 
```
 pip install git+https://github.com/executablebooks/jupyterlab-myst.git@agoose77/feat-jupyterlab-v4
```
which is a recipe that I've used successfully with the jupytext wip at least
it is a little slow but that's fine as long as it keeps the rather gory ts/jlpm/yarn details out of my way ;)

however in the case of `jupyterlab_myst` this fails with a long-ish error log, where the important part seems to be (full log in the details section below)
```
          message: 'Module build failed (from ./node_modules/css-loader/dist/cjs.js):\n' +
            "Error: Can't resolve 'app.css' in '/tmp/pip-req-build-hb92s9ma/style'\n" +
```

is this easily fixable ? 
it would be nice because it could help mere mortals like @fperez and myself more easily deploy and try it out

in any case, thanks for taking care of the upgrade 👍 

<details>
<summary>the end of the error log</summary>

<pre>
<code>
&lt;snip>
      ➤ YN0013: │ yargs-parser@npm:21.1.1 can't be found in the cache and will be fetched from the remote registry
      ➤ YN0013: │ yargs@npm:17.7.2 can't be found in the cache and will be fetched from the remote registry
      ➤ YN0013: │ yjs@npm:13.6.1 can't be found in the cache and will be fetched from the remote registry
      ➤ YN0013: │ yocto-queue@npm:0.1.0 can't be found in the cache and will be fetched from the remote registry
      ➤ YN0013: │ zwitch@npm:2.0.4 can't be found in the cache and will be fetched from the remote registry
      ➤ YN0000: └ Completed in 27s 544ms
      ➤ YN0000: ┌ Link step
      ➤ YN0007: │ @fortawesome/fontawesome-free@npm:5.15.4 must be built because it never has been before or the last one failed
      ➤ YN0000: └ Completed in 10s 724ms
      ➤ YN0000: Done with warnings in 38s 857ms
      INFO:hatch_jupyter_builder.utils:> /private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-build-env-tts1f8tl/overlay/bin/jlpm run build:prod
      Building extension in .

      Compilation starting…


      Compilation finished

      [
        {
          moduleIdentifier: '/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/css-loader/dist/cjs.js!/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/style/base.css',
          moduleName: './node_modules/css-loader/dist/cjs.js!./style/base.css',
          message: 'Module build failed (from ./node_modules/css-loader/dist/cjs.js):\n' +
            "Error: Can't resolve 'app.css' in '/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/style'\n" +
            '    at finishWithoutResolve (/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/enhanced-resolve/lib/Resolver.js:321:18)\n' +
            '    at /private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/enhanced-resolve/lib/Resolver.js:398:15\n' +
            '    at /private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/enhanced-resolve/lib/Resolver.js:447:5\n' +
            '    at eval (eval at create (/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:15:1)\n' +
            '    at /private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/enhanced-resolve/lib/Resolver.js:447:5\n' +
            '    at eval (eval at create (/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:27:1)\n' +
            '    at /private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/enhanced-resolve/lib/DescriptionFilePlugin.js:87:43\n' +
            '    at /private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/enhanced-resolve/lib/Resolver.js:447:5\n' +
            '    at eval (eval at create (/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:16:1)\n' +
            '    at /private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/enhanced-resolve/lib/forEachBail.js:18:13',
          moduleId: 41150,
          moduleTrace: [ [Object], [Object], [Object] ],
          details: undefined,
          stack: 'ModuleBuildError: Module build failed (from ./node_modules/css-loader/dist/cjs.js):\n' +
            "Error: Can't resolve 'app.css' in '/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/style'\n" +
            '    at finishWithoutResolve (/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/enhanced-resolve/lib/Resolver.js:321:18)\n' +
            '    at /private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/enhanced-resolve/lib/Resolver.js:398:15\n' +
            '    at /private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/enhanced-resolve/lib/Resolver.js:447:5\n' +
            '    at eval (eval at create (/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:15:1)\n' +
            '    at /private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/enhanced-resolve/lib/Resolver.js:447:5\n' +
            '    at eval (eval at create (/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:27:1)\n' +
            '    at /private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/enhanced-resolve/lib/DescriptionFilePlugin.js:87:43\n' +
            '    at /private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/enhanced-resolve/lib/Resolver.js:447:5\n' +
            '    at eval (eval at create (/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:16:1)\n' +
            '    at /private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/enhanced-resolve/lib/forEachBail.js:18:13\n' +
            '    at processResult (/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/webpack/lib/NormalModule.js:764:19)\n' +
            '    at /private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/webpack/lib/NormalModule.js:866:5\n' +
            '    at /private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/loader-runner/lib/LoaderRunner.js:400:11\n' +
            '    at /private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/loader-runner/lib/LoaderRunner.js:252:18\n' +
            '    at context.callback (/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/loader-runner/lib/LoaderRunner.js:124:13)\n' +
            '    at Object.loader (/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/css-loader/dist/index.js:123:5)\n' +
            '    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)',
          compilerPath: undefined
        }
      ]
      /private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-build-env-tts1f8tl/overlay/lib/python3.11/site-packages/jupyterlab/debuglog.py:56: UserWarning: An error occurred.
        warnings.warn("An error occurred.")
      /private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-build-env-tts1f8tl/overlay/lib/python3.11/site-packages/jupyterlab/debuglog.py:57: UserWarning: subprocess.CalledProcessError: Command '['node', '/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji/node_modules/@jupyterlab/builder/lib/build-labextension.js', '--core-path', '/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-build-env-tts1f8tl/overlay/lib/python3.11/site-packages/jupyterlab/staging', '/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-req-build-bnue2gji']' returned non-zero exit status 2.
        warnings.warn(msg[-1].strip())
      /private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-build-env-tts1f8tl/overlay/lib/python3.11/site-packages/jupyterlab/debuglog.py:58: UserWarning: See the log file for details: /var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/jupyterlab-debug-5hec7dl0.log
        warnings.warn(f"See the log file for details: {log_path!s}")
      Traceback (most recent call last):
        File "/Users/tparment/miniconda3/envs/ue12-p23-intro/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/Users/tparment/miniconda3/envs/ue12-p23-intro/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/Users/tparment/miniconda3/envs/ue12-p23-intro/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 152, in prepare_metadata_for_build_wheel
          whl_basename = backend.build_wheel(metadata_directory, config_settings)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-build-env-tts1f8tl/overlay/lib/python3.11/site-packages/hatchling/build.py", line 56, in build_wheel
          return os.path.basename(next(builder.build(wheel_directory, ['standard'])))
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-build-env-tts1f8tl/overlay/lib/python3.11/site-packages/hatchling/builders/plugin/interface.py", line 150, in build
          build_hook.initialize(version, build_data)
        File "/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-build-env-tts1f8tl/normal/lib/python3.11/site-packages/hatch_jupyter_builder/plugin.py", line 87, in initialize
          raise e
        File "/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-build-env-tts1f8tl/normal/lib/python3.11/site-packages/hatch_jupyter_builder/plugin.py", line 82, in initialize
          build_func(self.target_name, version, **build_kwargs)
        File "/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-build-env-tts1f8tl/normal/lib/python3.11/site-packages/hatch_jupyter_builder/utils.py", line 115, in npm_builder
          run([*npm_cmd, "run", build_cmd], cwd=str(abs_path))
        File "/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-build-env-tts1f8tl/normal/lib/python3.11/site-packages/hatch_jupyter_builder/utils.py", line 229, in run
          return subprocess.check_call(cmd, **kwargs)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/Users/tparment/miniconda3/envs/ue12-p23-intro/lib/python3.11/subprocess.py", line 413, in check_call
          raise CalledProcessError(retcode, cmd)
      subprocess.CalledProcessError: Command '['/private/var/folders/9n/sxs31qhj1gnd6gk2v0ns8848000fn2/T/pip-build-env-tts1f8tl/overlay/bin/jlpm', 'run', 'build:prod']' returned non-zero exit status 1.
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
</code>
</pre>
</details>

File: src/myst.ts [hunk 4]
```diff
@@ -62,26 +82,34 @@ export function markdownParse(text: string): Root {
   return mdast as Root;
```
Comment (agoose77) [inline src/myst.ts:82]: @rowanc1 I removed the change from https://github.com/executablebooks/jupyterlab-myst/pull/145 here, because I don't know how it fits into this PR, in which I've split the parsing into two routines. My hunch is the lifting logic needs to move from its current place. 

In any case, this PR is not the final API - I just want to roughly break-up the logic before tightening up the API.

Comment (rowanc1) [reply src/myst.ts:84]: Brought the relevant changes back!

File: .gitignore [hunk 5]
```diff
@@ -123,3 +123,6 @@ dmypy.json
 
 # Yarn cache
 .yarn/
+
+# Generated css files by tailwind
+style/app.css
```
Comment (rowanc1) [inline .gitignore]: @agoose77, this is a pre-built file, not sure why we lost it in the gitignore?

Comment (rowanc1) [reply .gitignore]: @parmentelat this is related to the error that you ran into. The app.css is built on demand, but apparently not being triggered on a pip install. I don't know quite enough about what pip is calling to be helpful though, hopefully we will get this in soon though!

File: package.json [hunk 6]
```diff
@@ -35,7 +35,7 @@
     "build:labextension:dev": "jupyter labextension build --development True .",
     "build:lib": "tsc --sourceMap",
     "build:lib:prod": "tsc",
-    "build:prod": "jlpm clean && jlpm build:lib:prod && jlpm build:labextension",
+    "build:prod": "jlpm clean && jlpm build:css && jlpm build:lib:prod && jlpm build:labextension",
```
Comment (rowanc1) [inline package.json:38]: @parmentelat I think this will actually fix the pip install.

**Comment by rowanc1** [comment]
Ok -- I think this is good enough to have a first major release of JupyterLab 4.0 support. There is still the double render, which is annoying but all of it works, and things are back up to date with where `main` was before. If #151 is merged then we also have basic backwards compatibility with 3.6 -- we could do better there, but I don't think it is worth it, maybe @agoose77 you have better ideas for how to do backwards compat on this new way of working?

**Comment by agoose77** [comment]
@rowanc1 i think we might be able to use much of this PR as-is on 3.6. At the very least, we can backport most of the code so that the two branches are similar, but it may be that we can use guards like #152!

**Comment by agoose77** [comment]
@rowanc1 i think we might be able to use much of this PR as-is on 3.6. At the very least, we can backport most of the code so that the two branches are similar, but it may be that we can use guards like #151!

**Comment by rowanc1** [comment]
please update playwright snapshots

**Comment by rowanc1** [comment]
I don't know why the update screenshots code is failing:
https://github.com/executablebooks/jupyterlab-myst/actions/runs/5237359362/jobs/9455487682

**Comment by rowanc1** [comment]
please update playwright snapshots

**Comment by rowanc1** [comment]
please update playwright snapshots

**Comment by rowanc1** [comment]
please update playwright snapshots

**Comment by rowanc1** [comment]
Turns out that the screenshot action is based off of main, and that is why there was a failure and the updates to the YAML were not working. Pretty unintuitive, but should be fine for anything except a jupyterlab release.

File: ui-tests/yarn.lock [hunk 7]
```diff
@@ -1,2841 +0,0 @@
-# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (rowanc1) [inline ui-tests/yarn.lock:1]: Not sure if we want to bring this yarn-lock back, or we could add it to the gitignore, it is probably fine for it to be generated every time.

**Review by rowanc1 — APPROVED** [review]
Final look through the code, I think all this is great and super happy to get this in and have a major release to support jlab 4. I think some of the things that happen right after that are the double render improvements and then try to get to some of the other major features that we need!

Nice work @agoose77 🚀

**Comment by parmentelat** [comment]
@rowanc1 

I do confirm the pip command now can carry out proper install 

So I could easily try out commit b40590343f24ed0ffc3f9363ba033b48dd454b51, and happily report that it works as expected, as far as I can tell for my relatively modest needs - primarily admonitions

thank you !

File: src/inlineExpression.tsx [hunk 8]
```diff
@@ -91,7 +91,7 @@ export class RenderedExpression extends Widget {
 }
 
 function PlainTextRenderer({ content }: { content: string }) {
-  content = content.replace(/^(["'])(.*)\1$/, '$2');
+  //content = content.replace(/^(["'])(.*)\1$/, '$2');
```
Comment (rowanc1) [inline src/inlineExpression.tsx:94]: Just browsing the most recent changes, this was in to remove quotes around strings. Is this change intentional?

**Review by rowanc1 — CHANGES_REQUESTED** [review]
@agoose77 -- tested this out, and some very strange behaviour on a subsequent rerender of a text-cell. Changing a first cell renders with the contents of the last cell.

**Comment by rowanc1** [comment]
Ok, fixed the wrong cell model being shown -- in my JupyterLab (or notebook?) the `cell.id` is an empty string. The object comparison works well here.

Right now there are a lot of errors being thrown, I think that these should be at most debug messages as we are actually expecting them to fail. Having this be the default state of the console is a bit rough - also not sure if we should leave the console logs in, but if we do, we should make sure that they are `debug` calls.

image

**Comment by agoose77** [comment]
Hmm, I've not been seeing that. In any case, I'm not happy about the PR as it stands - after our meeting I took another look and I have a hunch that we can revert back to something more like our original code (main). I know that we discussed releasing as is, but I've made good progress, I'll see tomorrow if it works and then we can feel more confident moving forward. I'll check out this bug as a part of that.

File: src/inlineExpression.tsx [hunk 9]
```diff
@@ -84,7 +83,7 @@ export class RenderedExpression extends Widget {
 
     // Create renderer
     const renderer = this.rendermime.createRenderer(mimeType);
-    layout.widget = renderer;
+    if (layout) layout.widget = renderer;
     console.assert(renderer.isAttached, 'renderer was not attached!', renderer);
```
Comment (rowanc1) [inline src/inlineExpression.tsx:87]: Many of these calls consistently fail, and aren't actually a problem. I think that we should not show console errors unless there is a problem. We could also put them behind a dev flag if we want.

**Review by rowanc1 — APPROVED** [review]
Ok, extension is working again with my minor changes! Without the double render is MUCH nicer from a UX perspective. Would prefer not to have the console errors when we release this, but isn't a big deal and can always be fixed later.

**Comment by rowanc1** [comment]
Not sure, maybe this will just go away .. says it is a github internal error:
image

File: src/index.ts [hunk 10]
```diff
@@ -138,7 +56,7 @@ const executorPlugin: JupyterFrontEndPlugin<void> = {
 };
 
 const mimeRendererPlugin: JupyterFrontEndPlugin<void> = {
-  id: 'jupyterlab-myst:mimeRenderer',
+  id: 'jupyterlab-myst:mime-renderer',
```
Comment (jtpio) [inline src/index.ts:59]: Does this plugin still need the optional dependency on `IMarkdownViewerTracker`?

It looks like it's not used within the plugin `activate` function.

Comment (jtpio) [reply src/index.ts:59]: Wondering if this could help with making the extension compatible with Notebook 7: https://github.com/jupyter/notebook/issues/6915

**Comment by fperez** [comment]
Thanks for the pip line @parmentelat! Indeed, for someone like me with such limited JS/TS expertise, being able to test like this is pretty important :) 

Here just to report that things are looking pretty good so far on a 4.0.2 installation, thanks so much @agoose77 @rowanc1!!

I am not sure if the math double rendering is entirely gone or just less noticeable due to performance improvements, but in any case the experience with inline math is much, much smoother and more reliable than very recently.  It was very hit or miss in my prior tests (I always had to demo it with a caveat to the audience :) and now it works pretty well (in light testing though).

I noticed that inline markdown links are not properly formatted, though they do work as URLs. See e.g. the example for the youtube link in the included example:

image

That was already probably on your radar, but just in case.

As usual thanks so much for the great work, LMK if I can be of any help with further testing/input. Very much looking forward to this being fully released!

File: src/index.ts (hunk 10 above)
Comment (agoose77) [reply src/index.ts:59]: We use this to ensure that JupyterLab has already registered the `text/markdown`: https://github.com/jupyterlab/jupyterlab/blob/d56ff811f39b3c10c6d8b6eb27a94624b753eb53/packages/markdownviewer-extension/src/index.ts#L71

It looks like you've already tackled this token being unavailable [upstream](https://github.com/jupyter/notebook/pull/6921), but let me know if we can do anything better here?

Comment (jtpio) [reply src/index.ts:59]: Right, some extensions use this trick to ensure plugin activation order.

Maybe we could add a small comment above the `optional` to document that behavior? 

> It looks like you've already tackled this token being unavailable [upstream](https://github.com/jupyter/notebook/pull/6921), but let me know if we can do anything better here?

That's right, `jupyterlab-myst` should now work in Notebook 7 without modifications.

**Comment by rowanc1** [comment]
Taken over by https://github.com/executablebooks/jupyterlab-myst/pull/155
